### ISSUE TITLE
PHP @var doesn't work

### ISSUE THREAD

**Issue by UziTech** [nikhilkalige/docblockr#255]
![image](https://cloud.githubusercontent.com/assets/97994/24463734/3c7b5ae2-146d-11e7-84c0-e0436d50ad99.png)

When creating a docblock for a variable it creates a docblock for the next function instead

**Comment by UziTech** [comment]
![image](https://cloud.githubusercontent.com/assets/97994/24464478/a8eb0ffe-146f-11e7-903c-f97d3137f2b1.png)

it looks like `parser.get_definition` doesn't stop after var definition

**Comment by UziTech** [comment]
We could setup a `.get_definition` function in `php.js` or a better solution might be to check which comes first `var` or `function` and use that information to call `parse_function` or `parse_var` in `DocsParser.prototype.parse`

**Comment by lxg** [comment]
This bug is a regression, because it did work before. (Can’t tell which version, though.)

**Comment by MoritzKn** [comment]
@UziTech It would be great if you could fix this and submit a PR. Ideally containing some specs to prevent future regression.

**Comment by UziTech** [comment]
I am planning on submitting a PR. I am just trying to get an idea of the best way to fix this. I just added a `get_definition` function to php.js and it works, but it seems a fix in the `DocsParser.prototype.parse` would be better but it might break other languages.

related to this issue:
starting a docblock before a class doesn't work because `DocsParser.prototype.parse` only looks for functions and vars

So how should this be fixed? by just patching php.js or by changing the things `.parse` looks for?

**Comment by MoritzKn** [comment]
The function `DocsParser.prototype.get_definition` is designed to generally work with C like languages. As long as there is nothing obviously buggy that affects all languages, I would recommend to implement `PhpParser.prototype.get_definition`.


> starting a docblock before a class doesn't work because DocsParser.prototype.parse only looks for functions and vars

What would you expect here? Docblockr will create a doc comment but no template. I think this is correct because classes normally don't need a template (no `@param`, etc...).

**Comment by UziTech** [comment]
> Docblockr will create a doc comment but no template.

No actually it will create a docblock comment for the first function because it ignores the class.

There is the exact save problem with javascript classes

![image](https://cloud.githubusercontent.com/assets/97994/24510972/418b99c0-1530-11e7-8762-00e487f7f9d6.png)

It seems like fixing php with a `PhpParser.prototype.get_definition` would just be putting a finger in the dam.

**Comment by MoritzKn** [comment]
Ok I see, in this case I would say there is something wrong with the `DocsParser.prototype.get_definition` function indeed.

**Comment by UziTech** [comment]
I've been trying to come up with the best way to fix this and I think the best way would be to include `matches.index` in the return value for `parse_function`, `parse_var`, and `parse_class` and in the `DocsParser.prototype.parse` function call all three and the one with the lowest index would be the closest to the doc comment and the one that should be used.

What do you think?

**Comment by MoritzKn** [comment]
i think `DocsParser.prototype.get_definition` should just stop after the first statement and only return this statement. If we try parsing every possibility, we run into problems if the comment is followed by a not statement with no specific parser (i.e. not a function or variable declaration). Take this for example.
```js
/**

if (true) {
    var foo = 42;
}
```
Hitting enter after `/**` in this case shouldn't add a template for `foo`.

**Comment by UziTech** [comment]
The only problem I could see with that is if a function definition spans multiple lines

```javascript
function test(
  $arg1,
  $arg2
) {
  return 0;
}
```

one thing we could try is to only get the definition up to a `;` or `{` this should be enough to get the definition. but it won't work for all languages.

Also in practice this can still pose a problem when a semi-colon is missed on a variable. I seem to forget them initially a lot.

**Comment by UziTech** [comment]
maybe could still get multiple lines with `get_definition`, remove any leading white space, and change `parse_*` to only match from the beginning

**Comment by MoritzKn** [comment]
The `get_definition` function already tries to do this and get the first statement, even if it spans multiple lines. But I think this is indeed the the core problem here -- finding the stat and end of a statement isn't trivial. Because of function statements like this `function (foo = {test: "default"})` only regexes won't work here either. I already thought about using atoms grammar scope for parsing, maybe this could make it easier.

**Comment by UziTech** [comment]
I like the idea of using grammar scope. 

Just make sure there are good tests in case the language package changes the scope output. I just ran into this issue on another package when `language-css` change their scopes from `punctuation.section.property-list.begin.css` to `punctuation.section.property-list.begin.bracket.curly.css`

**Comment by MoritzKn** [comment]
Yeah true. But it could be a real improvement. The current variant of just using regexes for every thing breaks on all ends. I think eventually the package has to be refactored / reworked in large parts anyway -- maybe even completely rewritten. I just don't have the time for it.

**Comment by MoritzKn** [comment]
Fixed by #259
